doc/cookbook/file-upload/FileUpload.lhs

```diff
@@ -0,0 +1,187 @@
+that one could use with `ReqBody`, but instead is its own
+dedicated `ReqBody`-like combinator named
+[`MultiPartForm`](https://hackage.haskell.org/package/servant-multipart-0.11/docs/Servant-Multipart.html#t:MultipartForm).
+
+This combinator takes two parameters. The first one is the
+"backend" to use. Currently, you only have the choice between
+`Mem` and `Tmp`. The former loads the entire input in memory,
+even the uploadedd files, while `Tmp` will stream uploaded
+files to some temporary directory.
+
+The second parameter is the type you want the multipart data
+to be decoded to. Indeed there is a `FromMultipart` class that
+allows you to specify how to decode multipart form data from
+`MultipartData` to a custom type of yours. Here we use the
+trivial "decoding" to `MultipartData` itself, and simply
+will get our hands on the raw input. If you want to use
+a type of yours, see the documentation for
+[`FromMultipart`](https://hackage.haskell.org/package/servant-multipart-0.11/docs/Servant-Multipart.html#t:FromMultipart).
+
+Our only request handler has type `MultipartData Mem -> Handler Integer`.
+All it does is list the textual and file inputs that
+were sent in the multipart request body. The textual
+inputs are in the `inputs` field while the file inputs
+are in the `files` field of `multipartData`.
+
+``` haskell
+-- MultipartData consists in textual inputs,
+-- accessible through its "inputs" field, as well
+-- as files, accessible through its "files" field.
+upload :: Server API
+upload multipartData = do
+  liftIO $ do
+    putStrLn "Inputs:"
+    forM_ (inputs multipartData) $ \input ->
+      putStrLn $ "  " ++ show (iName input)
+            ++ " -> " ++ show (iValue input)
+
+    forM_ (files multipartData) $ \file -> do
+      let content = fdPayload file
+      putStrLn $ "Content of " ++ show (fdFileName file)
+      LBS.putStr content
+  return 0
+
+startServer :: IO ()
+startServer = run 8080 (serve api upload)
+```
+
+Finally, a main function that brings up our server and
+sends some test request with `http-client` (and not
+servant-client this time, has servant-multipart does not
+yet have support for client generation.
+
+``` haskell
+main :: IO ()
+main = withSocketsDo . bracket (forkIO startServer) killThread $ \_threadid -> do
+  -- we fork the server in a separate thread and send a test
+  -- request to it from the main thread.
+  manager <- newManager defaultManagerSettings
+  req <- parseRequest "http://localhost:8080/"
+  resp <- flip httpLbs manager =<< formDataBody form req
+  print resp
+
+  where form =
+          [ partBS "title" "World"
+          , partBS "text" $ encodeUtf8 "Hello"
+          , partFileSource "file" "./README.md"
+          ]
+```
+
+If you run this, you should get:
+
+```
+$ cabal new-build cookbook-file-upload
+[...]
+$ dist-newstyle/build/x86_64-linux/ghc-8.2.1/cookbook-file-upload-0.1/x/cookbook-file-upload/build/cookbook-file-upload/cookbook-file-upload
```

Comment:
btw, `cabal new-run cookbook-file-upload` (it will also build the package, if not built already), given `cabal-install-head`.

Reply:
Oh, it's there now, good :)

Reply:
I'll wait for it to be released to use it in the tutorial though.